x-pack/plugins/infra/server/utils/create_date_range_aggregation/create_date_range_aggregation.ts

```diff
@@ -0,0 +1,255 @@
+  const options = Object.assign({}, optionsWithoutDefaults, defaults);
+  const { to, from } = normalizeTimerange(options.timerange);
+  const durationMs = to - from;
+
+  const rangeSettings = parseTargetRangeSize(options.targetRangeSize);
+  const rangeSizeMs = calculateRangeSize({
+    ...rangeSettings,
+    durationMs,
+    targetGranularity: options.targetGranularity,
+  });
+
+  if (durationMs < rangeSizeMs) {
+    // Is this always a problem, can I use the duration as a single sized bucket instead?
+    throw new Error('Duration is shorter than range size');
+  }
+
+  return generateDateRangeAggregation(
+    options.field,
+    { to, from },
+    durationMs,
+    rangeSizeMs,
+    options.partialRangeSetting
+  );
+}
+
+function normalizeTimerange(timerange: Timerange): NumberTimerange {
+  // How to account for timezones?
+  const { from, to } = timerange;
+
+  if (typeof from === 'string' && typeof to === 'string') {
+    if (from.includes('now') || from.includes('||') || to.includes('now') || to.includes('||')) {
+      const now = new Date();
+      const fromMoment = datemath.parse(from, { forceNow: now });
+      const toMoment = datemath.parse(to, { forceNow: now });
+
+      if (!fromMoment || !toMoment) {
+        throw new Error('Invalid datemath format');
+      }
+
+      return {
+        from: fromMoment.valueOf(),
+        to: toMoment.valueOf(),
+      };
+    }
+
+    const fromDate = new Date(from).getTime();
+    const toDate = new Date(to).getTime();
+
+    if (!isNaN(fromDate) && !isNaN(toDate)) {
+      return {
+        from: fromDate,
+        to: toDate,
+      };
+    }
+
+    throw new Error('Invalid date string format');
+  } else if (typeof from === 'number' && typeof to === 'number') {
+    const fromDate = new Date(from);
+    const toDate = new Date(to);
+
+    if (isNaN(fromDate.getTime()) || isNaN(toDate.getTime())) {
+      throw new Error('Invalid millisecond date range');
+    }
+
+    return timerange as NumberTimerange;
+  }
+
+  throw new Error('Unsupported timerange format');
+}
+
+function parseTargetRangeSize(targetRangeSize: string) {
+  if (targetRangeSize.startsWith('>=')) {
+    return {
+      rangeSizeMs: parseUnitString(targetRangeSize.substring(2)),
+      mode: 'minimum' as RangeSizeMode,
+    };
+  }
+
+  return {
+    rangeSizeMs: parseUnitString(targetRangeSize),
+    mode: 'exact' as RangeSizeMode,
+  };
+}
+
+function parseUnitString(valueUnitString: string) {
+  const valueUnitPattern = new RegExp(`^([\\d\\.]+)(${datemath.units.join('|')})$`);
+  const matches = valueUnitString.match(valueUnitPattern);
+
+  if (!matches) {
+    throw new Error(`Unsupported value unit format: ${valueUnitString}`);
+  }
+
+  const [_, value, unit] = matches;
+
+  if (unit === 'ms') {
+    return parseFloat(value);
+  } else {
+    return parseFloat(value) * unitToMilliseconds[unit];
+  }
+}
+
+function calculateRangeSize({
+  mode,
```

Comment:
The old code has a mode that is only bucket count (like 100 buckets for 20 minutes) but talking to Chris this seemed to be kind of an edge case because that path doesn't take into account the minimum metricset period, so you have to pair the `auto` interval with the `forceInterval` flag which in that case means something like "allow empty buckets".

In reflection it doesn't seem to be so often that we need exactly 100 buckets, and we can achieve a similar thing by using the `exact` mode and the minimum metricset period (or some sane default) which still allows for empty buckets.